ethcore/src/engines/authority_round/mod.rs

```diff
@@ -125,54 +140,96 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
 	}
 }
 
-// Helper for managing the step.
+/// Helper for managing the step.
 #[derive(Debug)]
 struct Step {
 	calibrate: bool, // whether calibration is enabled.
-	inner: AtomicUsize,
-	duration: u16,
+	inner: AtomicU64,
+	/// Planned durations of steps.
+	durations: BTreeMap<u64, u16>,
 }
 
 impl Step {
-	fn load(&self) -> u64 { self.inner.load(AtomicOrdering::SeqCst) as u64 }
+	fn load(&self) -> u64 { self.inner.load(AtomicOrdering::SeqCst) }
+
 	fn duration_remaining(&self) -> Duration {
+		self.opt_duration_remaining().unwrap_or_else(|| {
+			let ctr = self.load();
+			error!(target: "engine", "Step counter under- or overflow: {}, aborting", ctr);
+			panic!("step counter under- or overflow: {}", ctr)
+		})
+	}
+
+	fn opt_duration_remaining(&self) -> Option<Duration> {
 		let now = unix_now();
-		let expected_seconds = self.load()
-			.checked_add(1)
-			.and_then(|ctr| ctr.checked_mul(self.duration as u64))
-			.map(Duration::from_secs);
-
-		match expected_seconds {
-			Some(step_end) if step_end > now => step_end - now,
-			Some(_) => Duration::from_secs(0),
-			None => {
-				let ctr = self.load();
-				error!(target: "engine", "Step counter is too high: {}, aborting", ctr);
-				panic!("step counter is too high: {}", ctr)
-			},
+		let mut prev_dur = u64::from(self.durations[&0]);
+		let mut prev_step = 0u64;
+		let mut prev_time = 0u64;
+		let next_step = self.load().checked_add(1)?;
+		for (time, dur) in self.durations.iter().skip(1) {
+			let step_diff = time.checked_add(prev_dur)?.checked_sub(1)?.checked_sub(prev_time)?.checked_div(prev_dur)?;
+			let step = prev_step.checked_add(step_diff)?;
+			if step >= next_step {
+				break;
+			}
+			prev_step = step;
+			prev_time = step_diff.checked_mul(prev_dur)?.checked_add(prev_time)?;
+			prev_dur = u64::from(*dur);
+		}
+		let time = prev_time.checked_add(next_step.checked_sub(prev_step)?.checked_mul(prev_dur)?)?;
+		let step_end = Duration::from_secs(time);
+		if step_end > now {
```

Comment:
You can use `saturating_sub` instead of this `if-else`

Reply:
I think `Duration` has no `saturating_sub`. But we could convert `now` to seconds and back, if you prefer.